components/builder-graph/src/package_build_manifest_graph.rs

```diff
@@ -191,10 +201,59 @@ impl PackageBuildManifest {
         order
     }
 
+    /// Fixup for strict package build ordering.
+    // The execution ordering of the base graph is specified only by the direct dependencies of the
+    // package. However, our build workers don't use the dependencies to select what package to use,
+    // instead taking the latest package in the channel. This creates an antidependency (read
+    // before write) as it is possible that we build the next iteration of a package before all
+    // of the consumers of the last iteration have started; if that happens those packages might
+    // pick up the wrong iteration. This is likely harmless, except it makes the process
+    // nondeterministic and hard to debug. Constraining this will protect against this
+    // nondeterminism at the cost of some parallelism.
+    //
+    // To counter this, we will add extra dependencies to the graph. A package iteration n
+    // (InternalVersionedNode) will now have dependencies on all of the consumers of iteration n-1,
+    // guaranteeing they complete before it starts
+    //
+    // This fixup will not be necessary once we have build workers that can take exact dependencies.
+    //
+    pub fn constrain_package_cycles(&mut self) {
+        // Phase one: Identify all of the nodes needing constraint. This will be all
+        // InternalVersionedNode with version > 1
+        let mut fixup_targets = Vec::new();
+        for node in self.graph.nodes() {
+            match node {
+                UnresolvedPackageIdent::InternalVersionedNode(_, n) if n > 1 => {
+                    fixup_targets.push(node);
+                }
+                _ => (),
+            }
+        }
+
+        // Phase two: For each identified node, find the n-1 th version and make each package that
+        // depends on the n-1th node a dependency of the nth node.
+        for node in fixup_targets.iter() {
+            if let UnresolvedPackageIdent::InternalVersionedNode(ident, n) = node {
+                // always matches...
+                let prev_node = UnresolvedPackageIdent::InternalVersionedNode(*ident, n - 1);
+                // Modifying the graph while iterating over edges isn't ok.
+                let consumers: Vec<UnresolvedPackageIdent> =
+                    self.graph
+                        .neighbors_directed(prev_node, petgraph::EdgeDirection::Incoming)
+                        .collect();
+                for consumer in consumers {
+                    self.graph
+                        .add_edge(*node, consumer, EdgeType::ExternalConstraint);
+                }
+            }
+        }
+    }
```

Comment:
This is probably not an issue, however, I am curious about something.  Since we are "fixing up" the graph by adding dependencies, should we add any debugging messages to help identify how often and/or how many nodes we are adding? Otherwise, we may be making some assumptions about the graph content and to what extent our mutations are occurring. Intrigued if that would prove helpful or not.

Reply:
Interesting thought. A nitpick/nuance is that technically we're not adding dependencies, we're making some implicit relationships explicit. Don't mind adding a logging message, but it either will be a largish number anytime we have a cycle, or zero otherwise.